### PR TITLE
fix(screenshot): say when scale or rotation was not applied

### LINKED ISSUE #626: screenshot: `scale` and `rotation` are accepted and silently ignored on Chromium and tvOS (the warning never reaches the response)
## Summary

`screenshot` accepts `scale` and `rotation` on targets where neither is applied, returns a full-size
un-rotated PNG, and attaches **no note or warning** to the result.

### Chromium

`{scale: 0.5}` and `{rotation: "LandscapeLeft"}` both return the untouched 1600×1136 capture.

Root cause in the shipped `dist/tool-server.cjs` (~line 115658, `chromium-server/screenshot.ts`):
`tryLoadSharp()` swallows the require failure and `warnSharpMissingOnce()` writes the explanation to the
**tool-server's `process.stderr` only** — it never reaches the MCP response. `sharp` is not a dependency
of the published package (no `node_modules/sharp`, not in `package.json`), so this is every user's
default state rather than a local quirk. `scale` and `rotation` share the gate
(`DOWNSCALER_TO_KERNEL` / `ROTATION_DEGREES2` sit beside it).

### tvOS

`rotation: "Portrait"` vs `rotation: "LandscapeRight"` produce byte-identical output (md5 `9010b3dc…`
both). Defensible for a TV, but it is dropped silently.

## Impact

The documented "use a downscaled screenshot to save context" guidance does nothing on Chromium — every
Chromium screenshot costs full-resolution context — and a caller who asks for a rotation gets an image
that silently isn't rotated.

## Suggested fix

Surface the existing message as a `note` on the tool result whenever a geometry parameter is dropped.
That is a one-liner and would make the limitation self-describing.

## Related

#609 tracks the iOS side of screenshot orientation (different root cause: simulator-server is
orientation-unaware).

---
*Found during a full-surface QA sweep of **0.18.1** (`swmansion-argent-0.18.1.tgz`; the global install
is byte-identical to the tarball). Matrix: iPhone 17 Pro + iPhone 17 / iOS 26.5, Pixel_9 / Android 16
(API 36), Apple TV 4K / tvOS 26.5, Electron 42 (Chrome 148), Linux aarch64 (Lima) incl. the remote-link
leg, and the deviceless CLI. Host: macOS Darwin 25.5.0, Node v22.23.1.*

### PR DESCRIPTION
**Pull request by filip131311**
Fixes #626.

## Reproduced live (Electron testbed)

```
screenshot { scale: 0.25 }               → 1600×1136   (unscaled)
screenshot { rotation: "LandscapeLeft" } → 1600×1136   (unrotated)
```
Both returned nothing but the saved path.

`sharp` really is absent from the published package (no `node_modules/sharp`, not in `package.json`), so this is every user's default. And the existing stderr line is guarded by a **once-per-process** flag — after the first Chromium screenshot there was no signal *anywhere*, not even on stderr.

## The suggested one-liner wouldn't have worked

The issue proposes attaching a `note` to the tool result. I checked `argent-mcp/src/content.ts:75-83`: for an image-output tool the renderer returns **only** image blocks plus a `Saved: <path>` line and discards every other result field. A `note` on the result would have been silently dropped by the very client the issue is about.

**The delivery that does work is still tool-server-only.** `http.ts` already builds a response-envelope `note` (update reminders, active recordings) and has the tool's `data` in scope where it does it. So the tool returns a reserved key, `http.ts` pops it off and pushes it into that note — and `mcp-server.ts:369`, `argent-cli/run.ts:310` and `argent-tools-client` all render it **today**. No client change, and an older client works against a newer server immediately.

## The gate: explicitly asked **and** actually dropped

Both halves matter, and each alone is a bug:

- Gating on the parameter alone would fire on every visual-snapshot step — `flow-visual.ts` passes `scale: 1.0`, which is a no-op, not a loss. `rotation: "Portrait"` is likewise the identity.
- Gating on the effective value would report the `ARGENT_SCREENSHOT_SCALE` default the caller never asked for.

Verified live, all five cases:

| call | note |
|---|---|
| no geometry param (auto-screenshot shape) | — |
| `scale: 1.0` (visual-snapshot shape) | — |
| `rotation: "Portrait"` | — |
| `rotation: "LandscapeLeft"` | ✅ |
| `scale: 0.3` on Android (applied server-side) | — |

And `data` stays `{ image }` — the reserved key is stripped before it reaches the client, so `--json` output and non-image results are unchanged.

```
note: scale was not applied — this is the unmodified capture. Chromium image post-processing
needs the optional `sharp` package: run `npm install sharp` in the tool-server's environment
and retry, or work with the full-size image.
```

## Scope

The issue names Chromium and tvOS. **Vega has the identical gap** — `captureVegaScreenshotPng({ scale })` never receives rotation, same as `tvScreenshot(udid, scale, signal)` — so a Chromium-only fix would leave two silent droppers behind. The physical-iPhone runner capture (added on main later) has the same gap and is covered the same way. Both TV notes are worded so they don't read as retryable.

Two further drop sites covered while the plumbing was here:
- Chromium also loses `scale` when the PNG header can't be read — and that path's comment claimed a "falls back to sharp metadata" that the caller doesn't do. Comment corrected.
- `rotation`'s schema description promised Chromium rotation unconditionally while `scale` already carried the honest caveat.

## Deliberately not done

- **Implementing** rotation for tvOS/Vega. Both are cheap (`sips -r`; Vega already decodes with `pngjs`), but rotation on a fixed-landscape TV is close to meaningless — an agent asking for it is confused, and saying so is more useful than a sideways image.
- **Making `sharp` a real dependency** — ~30 MB of native binary for every consumer, most of whom never touch Chromium. `optionalDependencies` isn't a middle ground either (npm installs them by default). The real fix is to move the Chromium post-process onto the in-repo `pngjs` + `resizeDecodedPng` and delete the sharp branch — worth a follow-up, with the caveat that `resizeDecodedPng` is lanczos3-only, so the other three `downscaler` kernels would need writing or trimming.
- A **flow-nested** `screenshot` step won't surface the note, since the envelope is request-scoped. Accepted: a flow step passing explicit geometry to a TV/Chromium target is a corner case, and a result-field path can be layered on later without undoing this.

## Checks

- 3094 tests pass; 9 new. All 11 pre-existing screenshot tests pass unmodified.
- One unrelated flake on the first full run (`http-recording-note.test.ts`) that passed in isolation and on re-run — not from this change.
- SpiderShield 9.10; extract-tools 46/46; prettier, eslint, both typechecks clean; lock untouched.

## Docs

No docs update needed: `reference/tools.mdx` carries only a one-line row for `screenshot`, and the parameter-level caveats live in the tool description itself.

## Review follow-up (live-tested from the packed tarball, Electron + Apple TV sim)

- **Fixed a pre-existing squash on Chromium with `sharp` present:** `scale: 0.5 + rotation: LandscapeRight` returned 800×568 (rotated content forced into the unrotated aspect ratio) because sharp rotates before it resizes and the target was sized from the unrotated header with `fit: "fill"`. Now 568×800.
- "scale and rotation **were** not applied"; the TV note says "returned unrotated" (scale on the same call is still applied).
- `takeToolNote` strips the reserved key whatever its value.
- Added `http-tool-note.test.ts` for the envelope hoist itself (JSON + NDJSON, key stripped, empty value ignored).
- `bundle-tools.cjs` marks `sharp` external so a dev install of it no longer breaks the tool-server bundle.
- The `png-header-unreadable` corner case now gets honest wording: `sharp` is installed and a requested rotation still ran, so the note no longer claims an unmodified capture or prescribes `npm install sharp` (drop reason threaded through `MediaReady`).

---

## Revived 2026-09-17 (merged `main` at 2ed075602)

- #693 has merged; this now targets `main` directly. Conflicts in `http.ts` (main added the physical-iPhone signing-detection note in the same spot: both notes are kept, signing note first) and `screenshot/index.ts` (imports + the `rotation` description, where this PR's longer wording replaces main's one-liner).
- **New drop site from main:** the physical-iPhone path (#944) downscales but never rotates, and only the schema said "Ignored on physical iPhones". It now gets the same "cannot be transformed" note as Apple TV and Vega (0940692e4).
- Re-verified live on the merged branch: Electron testbed without `sharp`: `rotation: LandscapeLeft`, `scale: 0.25` and both together each return the note, while no param / `scale: 1` / `rotation: Portrait` return none, and `data` stays `{ image }`. The same calls on a `main` server return no note (bug still present). With `sharp` on `NODE_PATH`: 400×284, 1136×1600, and 568×800 for `scale 0.5 + LandscapeRight`, all without a note. Apple TV 4K sim: `LandscapeLeft + scale 0.5` → 1920×1080 plus the Apple TV note, and `Portrait` → no note. The physical-iPhone path was not live-tested because no phone was reachable.
- Checks: tool-server vitest 6335 passed, `typecheck:tests`, eslint, prettier, `npm run build`.

### CODE REVIEW

File: packages/tool-server/src/tools/screenshot/dropped-geometry.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * Chromium can do both transforms — it just needs `sharp`, which is optional
+ * and not shipped — so this case is worth telling the caller how to fix.
+ */
+export function chromiumDropNote(dropped: DroppedGeometry[]): string | undefined {
+  if (dropped.length === 0) return undefined;
+  return (
+    `${list(dropped)} was not applied — this is the unmodified capture. Chromium image ` +
+    `post-processing needs the optional \`sharp\` package: run \`npm install sharp\` in the ` +
+    `tool-server's environment and retry, or work with the full-size image.`
+  );
```
[the remark this answers is not in the document]
Comment (filip131311) [reply packages/tool-server/src/tools/screenshot/dropped-geometry.ts:64]: Both halves addressed: b3df89d5e fixes the verb ("scale and rotation were not applied"), and acf4dc43b makes the sharp instruction conditional — a `png-header-unreadable` drop now gets its own wording (rotation was applied, sharp already installed) instead of the install-sharp remedy.

File: packages/tool-server/src/tools/screenshot/index.ts [hunk 2]
```diff
@@ -27,7 +33,13 @@ const zodSchema = z.object({
...
     .optional()
     .describe(
-      "Orientation override for the screenshot (rotates the captured image after Page.captureScreenshot on Chromium). On Android the capture already follows the device's rotation."
+      "Orientation override. Rarely needed: on Android the capture already follows the device's " +
+        "current rotation, so it is upright without this. Setting it replaces that with a fixed " +
+        "rotation, which on a rotated device produces an image whose geometry no longer matches " +
+        "`describe` frames or gesture coordinates. On Chromium it rotates the captured image after " +
+        "Page.captureScreenshot, which requires the optional `sharp` dependency. Apple TV and Vega " +
+        "captures cannot be rotated at all. When a rotation is requested but not applied, the " +
+        "response carries a note saying so — the image is returned unrotated either way."
```
[the remark this answers is not in the document]
Comment (filip131311) [reply packages/tool-server/src/tools/screenshot/index.ts:42]: Added a "Docs" section to the PR description: no docs update needed — reference/tools.mdx carries only a one-line row for `screenshot`; the parameter-level caveats live in the tool description.

File: packages/tool-server/src/chromium-server/screenshot.ts [hunk 3]
```diff
@@ -107,6 +111,10 @@ export async function captureScreenshot(
             kernel: DOWNSCALER_TO_KERNEL[opts.downscaler ?? "lanczos3"],
             fit: "fill",
           });
```
[the remark this answers is not in the document]
Comment (filip131311) [reply packages/tool-server/src/chromium-server/screenshot.ts:122]: Confirmed live (Electron testbed, sharp installed): scale 0.5 + LandscapeRight returned 800×568 with the rotated content squashed. Fixed in b3df89d5e — the resize target is computed from the post-rotation dimensions (568×800 verified live for both quarter turns, 800×568 for PortraitUpsideDown).

File: packages/tool-server/src/chromium-server/screenshot.ts [hunk 4]
```diff
@@ -89,24 +89,39 @@ export async function captureScreenshot(
...
+          const targetW = Math.max(1, Math.round(rotatedW * scale));
+          const targetH = Math.max(1, Math.round(rotatedH * scale));
           pipeline = pipeline.resize(targetW, targetH, {
             kernel: DOWNSCALER_TO_KERNEL[opts.downscaler ?? "lanczos3"],
             fit: "fill",
           });
+        } else {
+          // Header unreadable: the rotation still runs, but the resize cannot
+          // be sized, so the scale the caller asked for is lost.
+          dropped.push("scale");
```
[the remark this answers is not in the document]
Comment (filip131311) [reply packages/tool-server/src/chromium-server/screenshot.ts:126]: Fixed in acf4dc43b: the drop reason now rides MediaReady, and the header-unreadable case says the requested rotation was still applied and sharp is already installed, instead of "unmodified capture" + "npm install sharp".
